### ISSUE TITLE
Remove 6 un-necessary package.json dependencies for new projects

### ISSUE THREAD

**Issue by devinrhode2** [solidjs/solid-start#989]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

FYI: Search "remove un-necessary deps vite esbuild" gave no results.

`npm init solid@latest` includes several deps that aren't needed for `npm run dev/build/start`

### Expected behavior 🤔

New projects' package.json files only include necessary deps

### Steps to reproduce 🕹

Steps:

1. `npm init solid@latest`
2. Remove these: esbuild, postcss, solid-js, @solidjs/router, @solidjs/meta, vite
3. Run `npm i`
4. Run `npm run dev/build/start`
5. Observe everything works without these deps.

### Context 🔦

New projects (with SSR and typescript) could have a package.json file like this:
```diff
     "start": "solid-start start"
   },
   "type": "module",
+  "optionalDependencies:comment": [
+    "esbuild uses `optionalDependencies` to install platform-specific binaries.",
+    "Using `npm install --omit=optional` will cause issues with esbuild.",
+    "Therefore, it's recommended to not use this package.json field."
+  ],
   "devDependencies": {
     "@types/node": "^18.16.19",
-    "esbuild": "^0.14.54",
-    "postcss": "^8.4.26",
     "solid-start-node": "^0.3.0",
-    "typescript": "^4.9.5",
-    "vite": "^4.4.6"
+    "typescript": "^4.9.5"
   },
+  "deps:comment": [
+    "These are needed for `npm run start` in production."
+  ],
   "dependencies": {
-    "@solidjs/meta": "^0.28.5",
-    "@solidjs/router": "^0.8.2",
-    "solid-js": "^1.7.9",
     "solid-start": "^0.3.1"
   },
   "engines": {
```

### Your environment 🌎

```shell
what command?
```

**Comment by MariuzM** [comment]
Hm not sure if this is a bug or more like feature request

**Comment by devinrhode2** [comment]
I'm using npm 9, bundled with Node 18.14,2

I guess we should add docs on production deployments.

I create a build outside docker, and then copy on just the dist folder. Pedantically, I have a `prod-package.json` which I copy in as `package.json`. It includes just `solid-start` and some "@suid/vite-plugin" used in my vite.config.ts, as `dependencies`. It includes a `"start": "solid-start start",` npm script along with `"type": "module",` and that's it.
> inside docker, I also copy in my package-lock.json, and my prod-package.json uses "*" semver specs for solid-start a @suid/vite-plugin. Npm doesn't pickup the latest when installing these two packages via `npm ci`. It only uses the version set in package-lock.json.

Imo `dependencies` should at least be used for deps needed to run production build (`solid-start start`).

Additionally, I think deps needed to _create_ a production build _could_ also be included.

If you're creating a library, then I think you should only have `devDeps`. I suppose ultimately there could be an extra question/option when running `create-solid`, "Are you creating a library?", etc.

I guess we could have a question on what package manager people are using... But who wants to write that code and maintain it? I think a universal package.json is better, so I can easily switch package managers.

**Comment by devinrhode2** [comment]
> For a library you probably have different requirements and define solid-start as a peerDep etc. There is work going on for a solid-cli. Maybe something to look at there?

I'm not sure, I'm not building a library (npm package) with Solid/SolidStart, I'll have to let others drive any efforts there.

> When building with the node adapter I would call `node dist/server.js` which is basically what the node adapter does by importing server.js.

I will certainly switch to doing this! This should alleviate the need for my `prod-package.json` :)

**Comment by devinrhode2** [comment]
One more note/question:
I noticed when `solid-start start` was running in my prod docker container, it didn't see any ui routes or api routes.

I COPY'd the `src` folder into the docker container, and viola, routes are being registered. It also seemed to be the case that we need esbuild to run the production server, because it's transpiling this `src` directory on the fly (hopefully with caching?).

I did explore moving my `npm run build` into docker, after having noticed some vite docs mentioning that vite will create some directories with pre-bundled deps. However, if I add a clean `node_modules` folder into git, and then create a build, nothing is added/changed in `node_modules`. I discovered my `.solid` directory is only related to the `vite-plugin-inspect` tool, which is driven by the `defineConfig({ inspect: boolean })` flag. Setting `inspect: false` leads to no `.solid` directory being created (or the directory is empty).

I'd imagine in the future, we will want to pre-compile the `src` folder, and avoid requiring esbuild to run a production server, working towards something like [Next.js's `standalone` feature](https://www.google.com/search?q=next.js+standalone).

**Comment by devinrhode2** [comment]
Overall, I think it'd be nice to document, and setup a contract, for production deployments.

There could even be a separate community-maintained Docker image and npm/shell scripts.

**Comment by edivados** [comment]
> One more note/question:
I noticed when solid-start start was running in my prod docker container, it didn't see any ui routes or api routes.

>I COPY'd the src folder into the docker container, and viola, routes are being registered. It also seemed to be the case that we need esbuild to run the production server, because it's transpiling this src directory on the fly (hopefully with caching?).

Yeah, the compiled server does not print any routes. Printing of the routes is done by the solid-start cli on the start command by using the router. This requires vite etc. things not needed in production. This is why I refer to the start command as preview build.

https://github.com/solidjs/solid-start/blob/c11e275114a386befcf51013674d48437a9a93a9/packages/start/bin.cjs#L471-L481

> However, if I add a clean node_modules folder into git, and then create a build, nothing is added/changed in node_modules

1. Don't add node_modules to git
2. The depdencies are in that case already there I guess

> I concluded that I don't need to run npm run build inside docker, because it doesn't pre-bundle anything, all I need is the dist folder inside docker.

No, you don't want to build inside your docker image or at least not on the final build stage if using [multi stage builds](https://docs.docker.com/build/building/multi-stage/).

**Comment by devinrhode2** [comment]
> 1. Don't add node_modules to git
yes 🤣 - I was only doing this to really double check if `solid-start build` is adding anything to a `node_modules/.cache` or `node_modules/.vite` directory.

**Comment by edivados** [comment]
Oh got it. I also read `npm install` for some reason and focused on the package installation. 😄

**Comment by devinrhode2** [comment]
> Yeah, the compiled server does not print any routes. Printing of the routes is done by the solid-start cli on the start command by using the router. This requires vite etc. things not needed in production. This is why I refer to the start command as preview build.

So you are saying, that the server works just fine, but it simply doesn't print out the routes?

**Comment by edivados** [comment]
Yes, exactly.
